### DISCUSSION TITLE
Transpiling async breaks way to get AsyncFunction constructor

### DISCUSSION THREAD

**Discussion by jayjun** [alpinejs/alpine#2318] [7. Bugs (archived)] [answered]
Problem described in detail here https://github.com/evanw/esbuild/issues/1743.

In summary, esbuild transpiles the `async` function inside
https://github.com/alpinejs/alpine/blob/9146daa567aece18e87f8a259445d54ce8e7341e/packages/alpinejs/src/evaluator.js#L54

into a regular function, thereby making `AsyncFunction` a regular function constructor. This breaks code that expects a `Promise` after using `AsyncFunction` as a constructor.

https://github.com/alpinejs/alpine/blob/9146daa567aece18e87f8a259445d54ce8e7341e/packages/alpinejs/src/evaluator.js#L94

Firefox Developer Tools:
```
Error: func(...).catch is not a function
```

I know it’s not directly Alpine’s fault. But I don’t see why we cannot use vanilla promises instead of this weird (but documented by [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncFunction)) way to dynamically create an async function.

Apparently `tsc` and `babel` cannot handle this too so changing to promises will increase compatibility.

**Comment by danddanddand** [comment]
Can you provide example of the expression you are getting the `Error: func(...).catch is not a function` from or reproduction steps?

**Comment by jayjun** [comment, marked as answer]
Edit: Since Alpine v3 supports `await` in expressions, it can only be used in environments with async support. See discussion in comments.

Sure, this error is triggered by any expression actually. To reproduce, Alpine must be transpiled by a bundler that transforms `async`, e.g. esbuild with `es2016` target.

### Steps to reproduce

1. Install Alpine.js and esbuild.

```
npm install alpinejs esbuild
```

2. Create `src.js`.

```javascript
import Alpine from "alpinejs";

Alpine.start();
```

3. Bundle into `dist.js`.

```
./node_modules/.bin/esbuild --bundle src.js --outfile=dist.js --target=es2016
```

3. Create minimal `index.html`.

```html
<html>
  <head>
    <script defer src="dist.js"></script>
  </head>
  <body x-data="{}">
  </body>
</html>
```

4. Open `index.html` in a browser.

**Reply by danddanddand** [reply]
Okay, I got replication and played with it a bit. So esbuild with newer target (esnext or es2020) seems to work fine. I use Snowpack for development and Webpack for production build and both seem to work fine. 

I can't see Alpine way to resolve this without changing evaluator to not be async (I think petite-vue uses this strategy https://github.com/vuejs/petite-vue/blob/main/src/eval.ts). 

Is there reason you need to target es2016 with esbuild?

**Reply by jayjun** [reply]
Not really, `es2016` is just the default generated by [Phoenix](https://github.com/phoenixframework/phoenix/blob/ff9fc80e5c646bf787baed1d09d59e8905dc09eb/installer/templates/phx_single/config/config.exs#L39).

As rightly pointed here https://github.com/evanw/esbuild/issues/1743#issuecomment-961159790, I shouldn’t expect Alpine to work in environments without async support. Is async really required though?

If changed to bare promises, we can keep the same `then` and `catch` code (presumably for bad expressions) so it’s not as barebones as petite-vue.

**Reply by danddanddand** [reply]
The AsyncFunction is used to do evaluation/maintain scope so I don't understand what you mean by "change to bare promises" as that wouldn't provide a way to do evaluation which is central to how Alpine works and usage in this instance.

Seems like updating target for esbuild is only solution. 

I assume the advantage of using AsyncFunction is mostly for speed when initializing and being able to run asynchronous code in expressions so I don't expect it would change in v3.

**Reply by jayjun** [reply]
Right, I didn’t know v3 supports `await` in expressions and this is how it’s implemented. All good then, thanks for the discussion!

**Reply by SimoTod** [reply]
Async is ecma2017. I assume phoenix targets 2016 by default to support old browser like internet explorer but Alpine v3 dropped the support for those browsers.

**Reply by jbosse** [reply]
Changing `--target=es2016` to `--target=es2017` in the Phoenix `config.exs` fixed my issue (for anyone like me trying to find where in the dependency chain things went awry)
